### PR TITLE
foldcolumn autowidth

### PR DESCRIPTION
**Pull request by teto**
like signcolumn, let foldcolumn adapt its width depending on the number of folds.
For instance `set fdc=auto:3`. It is backwards compatible in the sense that `set fdc=4` works but to simplify the code, I removed supportd for width > 9 (currently foldcolumn can be set to a max of 12). Can be supported if necessary.

WIP since I haven't any added any test yet (and some functionaltest still failing).

### CODE REVIEW

**Comment by teto** [comment]
finally marking this as RFC, I will be using as my neovim to check for bugs but iit should be fine, I've added tests etc.

Once merged, I will close https://github.com/neovim/neovim/pull/6073

File: src/nvim/ex_cmds.c
Comment (justinmk) [inline src/nvim/ex_cmds.c]: might as well keep the doc, the parallel form is helpful.

```suggestion
      set_string_option_direct((char_u *)"fdc", -1,  // no 'foldcolumn'
```

File: src/nvim/option.c
Comment (justinmk) [inline src/nvim/option.c]: unnecessary blank line

File: src/nvim/options.lua
Comment (justinmk) [inline src/nvim/options.lua]: it pains me, but ....... this ... is going to break some plugin. Perhaps we can merge this on master and see if it breaks anything that actually matters.

File: src/nvim/window.c
Comment (justinmk) [inline src/nvim/window.c]:
```suggestion
    int needed_fdccols = getDeepestNesting(wp);
```

File: src/nvim/window.c
Comment (justinmk) [inline src/nvim/window.c]: what is "folds signs"?

File: src/nvim/window.c
Comment (justinmk) [inline src/nvim/window.c]: "maximum" is a pretty ceremonious name, why not just "max"? (Only mentioning this for future reference...)

**Comment by justinmk** [comment]
Awesome idea! cc @da-x

- Can you update the docs similar to `:help 'signcolumn'`
- A before/after screenshot would be helpful for newsletter

File: src/nvim/options.lua
Comment (teto) [reply src/nvim/options.lua]: I know, even from a code point of view, having an other option "autofold" would have been easier to implement (instead of dealing with string allocation) but I wanted to remain consistent with the signcolumn syntax.

**Comment by teto** [comment]
sigh I had written the doc but messed up while squashing the code,  I put it back.

As for the screenshot, what do you mean ? This won't be visible except in a gif mqybe but I am not sure how to recortd the animation (https://github.com/phw/peek maybe, any recommendation ?)
2020-03-01-162544_461x145_scrot.

What do you think of the limitation from 12 to 9 in the maximum number of folds ? should we just push it and fix it later in case of complaints ?

File: src/nvim/window.c
Comment (teto) [reply src/nvim/window.c]: done

File: src/nvim/window.c
Comment (teto) [reply src/nvim/window.c]: I removed it altogether

File: runtime/doc/fold.txt
Comment (teto) [inline runtime/doc/fold.txt]: I was not sure how to edit this paragraph, the mentions to '+' and '|' should now refer to fillchars foldclose/foldopen but this made the text a bit verbose.

File: src/nvim/options.lua
Comment (justinmk) [reply src/nvim/options.lua]: Yeah, I'm glad you did it this way.

**Comment by justinmk** [comment]
> What do you think of the limitation from 12 to 9 in the maximum number of folds ? should we just push it and fix it later in case of complaints ?

Yeah, maybe document it if it can be done in <=10 words...

**Comment by justinmk** [comment]
> What do you think of the limitation from 12 to 9 in the maximum number of folds ? should we just push it and fix it later in case of complaints ?

9 should be fine. I don't see how 3 more fold columns would be important.

File: runtime/doc/fold.txt
Comment (justinmk) [reply runtime/doc/fold.txt]: > the mentions to '+' and '|' should now refer to fillchars foldclose/foldopen but this made the text a bit verbose.

Referring to  '+' and '|' is intuitive enough IMO.

File: runtime/doc/vim_diff.txt
Comment (justinmk) [inline runtime/doc/vim_diff.txt]: Is this true? Looks like it is "0" in options.lua. https://github.com/neovim/neovim/pull/11716/files#diff-2b89728238835a06b3522ac53e43510bR838

And it should probably stay "0", we don't have a strong reason to change it.

**Review by justinmk — APPROVED** [review]

**Comment by teto** [comment]
I did not have any crash after numerous use and it works as intended (diffmode too) so lets merge

**Comment by h-michael** [comment]
@justinmk 
This is a breaking change, right?
What about putting a breaking change label on such PRs for users and plugin developers?

**Comment by justinmk** [comment]
> What about putting a breaking change label on such PRs

I use the `compatibility` tag for that, loosely. Probably a new tag is a good idea, but users should look at `vim_diff.txt` and https://github.com/neovim/neovim/wiki/Following-HEAD for breaking changes. Not issues. 

This is technically a breaking change but it might not break any actual plugins, because:

> Vimscript casts to/from string/number so it might be ok.

 Waiting to see reports of actual breakage.

**Comment by Shougo** [comment]
denite.nvim breaks by it.
It set foldcolumn by Python.
So it is not converted automatically.
I will fix it later.

**Comment by h-michael** [comment]
@justinmk 
> I use the compatibility tag for that, loosely. Probably a new tag is a good idea, but users should look at vim_diff.txt and https://github.com/neovim/neovim/wiki/Following-HEAD for breaking changes. Not issues.

I see. thanks :)

**Comment by teto** [comment]
Sorry for breakage; I had mentioned the change after the merge at https://github.com/neovim/neovim/wiki/Following-HEAD#20191223 but good to know about the compatibility flag

**Comment by chengzeyi** [comment]
When editing large files, setting 'foldcolumn' to auto makes neovim work extremely slow. I suspect the current implementation  has significant performance issues.

**Comment by Asheq** [comment]
@teto I was pleasantly surprised to stumble onto this today. Just wanted to drop a thanks! 🙏

**Pull request merged**
